test-framework/munit/src/main/scala/com/dimafeng/testcontainers/TestContainersSuite.scala

```diff
@@ -0,0 +1,131 @@
+import org.testcontainers.lifecycle.TestDescription
+
+trait TestContainersSuite { self: Suite =>
+  /**
+   * To use testcontainers suites you need to declare,
+   * which containers you want to use inside your tests.
+   *
+   * For example:
+   * {{{
+   *   override type Containers = MySQLContainer
+   * }}}
+   *
+   * If you want to use multiple containers inside your tests, use `and` syntax:
+   * {{{
+   *   override type Containers = MySQLContainer and PostgreSQLContainer
+   * }}}
+   */
+  type Containers <: Andable
+
+  /**
+   * Contains containers startup logic.
+   * In this method you can use any intermediate logic.
+   * You can pass parameters between containers, for example:
+   * {{{
+   * override def startContainers(): Containers = {
+   *   val container1 = Container1.Def().start()
+   *   val container2 = Container2.Def(container1.someParam).start()
+   *   container1 and container2
+   * }
+   * }}}
+   *
+   * @return Started containers
+   */
+  def startContainers(): Containers
+
+  /**
+   * To use containers inside your test bodies you need to use `withContainers` function:
+   * {{{
+   * test("test") {
+   *   withContainers { mysqlContainer =>
+   *     // Inside your test body you can do with your container whatever you want to
+   *     assert(mysqlContainer.jdbcUrl.nonEmpty)
+   *   }
+   * }
+   * }}}
+   *
+   * `withContainers` also supports multiple containers:
+   * {{{
+   * test("test") {
+   *   withContainers { case mysqlContainer and pgContainer =>
+   *     // test body
+   *   }
+   * }
+   * }}}
+   *
+   * @param runTest Test body
+   */
+  def withContainers[A](runTest: Containers => A): A = {
+    val c = startedContainers.getOrElse(throw IllegalWithContainersCall())
+    runTest(c)
+  }
+
+  /**
+   * Override, if you want to do something after containers start.
+   */
+  def afterContainersStart(containers: Containers): Unit = {}
+
+  /**
+   * Override, if you want to do something before containers stop.
+   */
+  def beforeContainersStop(containers: Containers): Unit = {}
+
+  @volatile private[testcontainers] var startedContainers: Option[Containers] = None
+
+  val suiteDescription = TestContainersSuite.createDescription(self)
+
+  private[testcontainers] def beforeTest(containers: Containers): Unit = {
+    containers.foreach {
+      case container: TestLifecycleAware => container.beforeTest(suiteDescription)
+      case _ => // do nothing
+    }
+  }
+
+  private[testcontainers] def afterTest(containers: Containers, throwable: Option[Throwable]): Unit = {
+    containers.foreach {
+      case container: TestLifecycleAware => container.afterTest(suiteDescription, throwable)
+      case _ => // do nothing
+    }
+  }
+
+  private[testcontainers] def stopContainers(containers: Containers): Unit = {
+    try {
+      beforeContainersStop(containers)
+    }
+    finally {
+      try {
+        startedContainers.foreach(_.stop())
+      }
+      finally {
+        startedContainers = None
```

Comment:
Correct me if im wrong, but it looks like you have a potential race condition here. If you add a new container between line 103 and line 106 are executed (and those lines are not atomic), you wont stop the new container. From my understanding all operations on `containers` should be synchronized

Reply:
Actually I'm not sure it's worth to change state of `startedContainers` here at all. Probably at this point, it doesn't change anything.